libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -0,0 +1,225 @@
+-- The Gen Error monad
+------------------------------------------------------------------------
+
+-- | A class for different types of errors with a stack of `explain` calls to
+-- narrow down problems.
+class Monad m => MonadGenError m where
+  genError :: HasCallStack => [String] -> m a
+  fatalError :: HasCallStack => [String] -> m a
+  explain :: HasCallStack => [String] -> m a -> m a
+
+-- | The Gen Error monad, distinguishes between fatal errors
+-- and non-fatal errors.
+data GE a
+  = FatalError [[String]] [String]
+  | GenError [[String]] [String]
+  | Result [[String]] a
+  deriving (Ord, Eq, Show, Functor)
+
+instance Applicative GE where
+  pure = Result []
+  (<*>) = ap
+
+instance Monad GE where
+  FatalError es err >>= _ = FatalError es err
+  GenError es err >>= _ = GenError es err
+  Result _ a >>= k = k a
+
+instance Alternative GE where
+  empty = GenError [] []
+  m@FatalError {} <|> _ = m
+  GenError {} <|> m = m
+  Result es x <|> _ = Result es x
+
+instance MonadGenError GE where
+  genError = GenError []
+  fatalError = FatalError []
+  explain es (GenError es' err) = GenError (es : es') err
+  explain es (FatalError es' err) = FatalError (es : es') err
+  explain es (Result es' a) = Result (es : es') a
+
+instance MonadGenError m => MonadGenError (GenT m) where
+  genError es = GenT $ \_ -> pure $ genError es
+  fatalError es = GenT $ \_ -> pure $ fatalError es
+  explain es gen = GenT $ \mode -> fmap (explain es) (runGenT gen mode)
+
+instance MonadGenError m => MonadFail (GenT m) where
+  fail = genError . (: [])
+
+catGEs :: MonadGenError m => [GE a] -> m [a]
+catGEs [] = pure []
+catGEs (Result _ a : ges) = (a :) <$> catGEs ges
+catGEs (GenError {} : ges) = catGEs ges
+catGEs (FatalError es e : _) =
+  runGE $ FatalError es e
+
+fromGE :: ([String] -> a) -> GE a -> a
+fromGE _ (Result _ a) = a
+fromGE a (GenError es e) = a $ concat es ++ e
+fromGE _ (FatalError es e) =
+  error . unlines $ concat es ++ e
+
+errorGE :: GE a -> a
+errorGE = fromGE (error . unlines)
+
+isOk :: GE a -> Bool
+isOk GenError {} = False
+isOk FatalError {} = False
+isOk Result {} = True
+
+runGE :: MonadGenError m => GE r -> m r
+runGE (GenError es err) = foldr explain (genError err) es
+runGE (FatalError es err) = foldr explain (fatalError err) es
+runGE (Result es a) = foldr explain (pure a) es
+
+fromGEProp :: Testable p => GE p -> Property
+fromGEProp (GenError es err) = foldr (counterexample . unlines) (counterexample (unlines err) False) es
+fromGEProp (FatalError es err) = foldr (counterexample . unlines) (counterexample (unlines $ "Fatal error: " : err) False) es
+fromGEProp (Result es p) = foldr (counterexample . unlines) (property p) es
+
+fromGEDiscard :: Testable p => GE p -> Property
+fromGEDiscard (Result es p) = foldr (counterexample . unlines) (property p) es
+fromGEDiscard _ = discard
+
+headGE :: Foldable t => t a -> GE a
+headGE t
+  | null t = fatalError ["head of empty structure"]
+  | otherwise = pure $ head $ toList t
```

Comment:
Could you please rewrite this without using `head`, because ghc-9.8 will complain with a warning